### PR TITLE
Allow c++ ComponentBatch to be constructed for Vec of types that can be used to construct the inner component

### PR DESCRIPTION
**Pull request by jleibs**
### What

As an alternative (or in addition to) https://github.com/rerun-io/rerun/pull/3967 this introduces a pattern wherein any `ComponentBatch<C>` can be constructed from any `Vec<T>` where we have implemented an extension for constructing `C` from `T`. 

Additionally, this introduces a similar extension for constructing linestrips from vecs of Vec3-like things so that users don't event need to pick up a Vec3 dependency if they don't want to.

These two combine to allow:
```
    std::vector<std::vector<std::array<float, 3>>> points = {
        {{0.f, 0.f, 0.f}, {0.f, 0.f, 1.f}},
        {{1.f, 0.f, 0.f}, {1.f, 0.f, 1.f}},
        {{1.f, 1.f, 0.f}, {1.f, 1.f, 1.f}},
        {{0.f, 1.f, 0.f}, {0.f, 1.f, 1.f}},
    };

    rec.log("segments", rerun::LineStrips3D(points));
```

To do the correct thing.

I think this is important, relative to the variadic / initializer-list approach since it's more likely to match the kind of patterns users will encounter in the real-world, where they want to build a Vec of Vecs without needing to instantiate rerun types directly.

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/{{ pr.number }}) (if applicable)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/{{ pr.number }})
- [Docs preview](https://rerun.io/preview/{{ pr.commit }}/docs) 
- [Examples preview](https://rerun.io/preview/{{ pr.commit }}/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

**Comment by nikolausWest** [comment]
This looks nice!

File: docs/code-examples/line_segments2d_simple.cpp
Comment (emilk) [inline docs/code-examples/line_segments2d_simple.cpp:9]:
```suggestion
    std::vector<std::vector<std::array<float, 2>>> points = {
```

**Review by emilk — APPROVED** [review]

File: rerun_cpp/src/rerun/component_batch.hpp
Comment (Wumpf) [inline rerun_cpp/src/rerun/component_batch.hpp:323]: I think I really really want this. But nonetheless two concerns:
* I believe this makes error messages much much worse: If you now have a vector of things that can't be converted, it will end up in this specialization and fail there
   * have to try this
   * maybe we can to an std::enable_if on constructability to fix this!
* This always does a copy even when we do have binary compatibility.
   * worse still I'm not even sure if we may hit this path if T==TComponent. Again maybe an std::enable_if to be sure of that?
   * I don't know a quick and streamlined path out of the general copy here since we _want_ it for many cases. Imho 'usability > potential copy' though so maybe not that much of a concern

File: rerun_cpp/src/rerun/components/line_strip2d_ext.cpp
Comment (Wumpf) [inline rerun_cpp/src/rerun/components/line_strip2d_ext.cpp:18]: I think we might be able to get rid of this in the very near future by having `LineStrip2D` hold a `rerun::Collection` 🤔

**Review by Wumpf — CHANGES_REQUESTED** [review]
It makes a lot of sense and I want it.
Also love that you were able to implement this via an Adapter and not via a constructor on the batch as previusly suggested via DM
Just let me check if I can alleviate the short cuts I pointed out :)

**Comment by Wumpf** [comment]
Looks like I was correct on it breaking error messages. We no longer hit the static assert I wrote for 
```
    std::vector<std::string> strings = {"a", "b", "c"};
    rerun::ComponentBatch<Position2D> batch(strings);
```

Instead you get now 
```
/usr/bin/clang++ -DLOGURU_STACKTRACES=1 -I/Users/andreas/dev/rerun-io/rerun/build/_deps/logurugitrepo-src -I/Users/andreas/dev/rerun-io/rerun/build/_deps/catch2-src/src/catch2/.. -I/Users/andreas/dev/rerun-io/rerun/build/_deps/catch2-build/generated-includes -I/Users/andreas/dev/rerun-io/rerun/rerun_cpp/src -isystem /opt/homebrew/include -g -std=gnu++17 -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk -mmacosx-version-min=11.0 -Wall -Wcast-align -Wcast-qual -Wextra -Wformat=2 -Wmissing-include-dirs -Wnull-dereference -Wold-style-cast -Wpedantic -Wpointer-arith -Wshadow -Wswitch-enum -Wvla -Wc++17-compat-pedantic -Wc99-extensions -Wgnu -Wnon-gcc -Wshadow-all -g -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -fno-optimize-sibling-calls -fsanitize=address -MD -MT rerun_cpp/tests/CMakeFiles/rerun_sdk_tests.dir/component_batch.cpp.o -MF rerun_cpp/tests/CMakeFiles/rerun_sdk_tests.dir/component_batch.cpp.o.d -o rerun_cpp/tests/CMakeFiles/rerun_sdk_tests.dir/component_batch.cpp.o -c /Users/andreas/dev/rerun-io/rerun/rerun_cpp/tests/component_batch.cpp
In file included from /Users/andreas/dev/rerun-io/rerun/rerun_cpp/tests/component_batch.cpp:3:
In file included from /Users/andreas/dev/rerun-io/rerun/rerun_cpp/src/rerun/archetypes/points2d.hpp:6:
/Users/andreas/dev/rerun-io/rerun/rerun_cpp/src/rerun/archetypes/../component_batch.hpp:305:24: error: no matching conversion for functional-style cast from 'const std::string' to 'rerun::components::Position2D'
                return TComponent(datum);
                       ^~~~~~~~~~~~~~~~
/Users/andreas/dev/rerun-io/rerun/rerun_cpp/src/rerun/archetypes/../component_batch.hpp:101:52: note: in instantiation of member function 'rerun::ComponentBatchAdapter<rerun::components::Position2D, std::vector<std::string>>::operator()' requested here
        ComponentBatch(T&& input) : ComponentBatch(TAdapter<T>()(std::forward<T>(input))) {}
                                                   ^
/Users/andreas/dev/rerun-io/rerun/rerun_cpp/tests/component_batch.cpp:234:39: note: in instantiation of function template specialization 'rerun::ComponentBatch<rerun::components::Position2D>::ComponentBatch<std::vector<std::string> &>' requested here
    rerun::ComponentBatch<Position2D> batch(strings);
                                      ^
/Users/andreas/dev/rerun-io/rerun/rerun_cpp/src/rerun/archetypes/../components/position2d.hpp:23:16: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const std::string' to 'const rerun::components::Position2D' for 1st argument
        struct Position2D {
               ^
/Users/andreas/dev/rerun-io/rerun/rerun_cpp/src/rerun/archetypes/../components/position2d.hpp:23:16: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const std::string' to 'rerun::components::Position2D' for 1st argument
/Users/andreas/dev/rerun-io/rerun/rerun_cpp/src/rerun/archetypes/../components/position2d.hpp:46:13: note: candidate constructor not viable: no known conversion from 'const std::string' to 'rerun::datatypes::Vec2D' for 1st argument
            Position2D(rerun::datatypes::Vec2D xy_) : xy(xy_) {}
            ^
/Users/andreas/dev/rerun-io/rerun/rerun_cpp/src/rerun/archetypes/../components/position2d.hpp:53:13: note: candidate constructor not viable: no known conversion from 'const std::string' to 'std::array<float, 2>' for 1st argument
            Position2D(std::array<float, 2> xy_) : xy(xy_) {}
            ^
/Users/andreas/dev/rerun-io/rerun/rerun_cpp/src/rerun/archetypes/../components/position2d.hpp:44:13: note: candidate constructor not viable: requires 0 arguments, but 1 was provided
            Position2D() = default;
            ^
/Users/andreas/dev/rerun-io/rerun/rerun_cpp/src/rerun/archetypes/../components/position2d.hpp:33:13: note: candidate constructor not viable: requires 2 arguments, but 1 was provided
            Position2D(float x, float y) : xy{x, y} {}
            ^
1 error generated.
```


I'll try fixing that

**Comment by Wumpf** [comment]
oh my god. This was hard. But I figured it out. It even makes sense when you squint 😄 
The nice thing is that this solution is pretty powerful in general and preserves the error message. It's just a liiittle bit of personal defeat that it's SFINAE based, but as long as the error is sane who am I to complain:

```
/usr/bin/clang++ -DLOGURU_STACKTRACES=1 -I/Users/andreas/dev/rerun-io/rerun/build/_deps/logurugitrepo-src -I/Users/andreas/dev/rerun-io/rerun/build/_deps/catch2-src/src/catch2/.. -I/Users/andreas/dev/rerun-io/rerun/build/_deps/catch2-build/generated-includes -I/Users/andreas/dev/rerun-io/rerun/rerun_cpp/src -isystem /opt/homebrew/include -g -std=gnu++17 -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk -mmacosx-version-min=11.0 -Wall -Wcast-align -Wcast-qual -Wextra -Wformat=2 -Wmissing-include-dirs -Wnull-dereference -Wold-style-cast -Wpedantic -Wpointer-arith -Wshadow -Wswitch-enum -Wvla -Wc++17-compat-pedantic -Wc99-extensions -Wgnu -Wnon-gcc -Wshadow-all -g -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -fno-optimize-sibling-calls -fsanitize=address -MD -MT rerun_cpp/tests/CMakeFiles/rerun_sdk_tests.dir/component_batch.cpp.o -MF rerun_cpp/tests/CMakeFiles/rerun_sdk_tests.dir/component_batch.cpp.o.d -o rerun_cpp/tests/CMakeFiles/rerun_sdk_tests.dir/component_batch.cpp.o -c /Users/andreas/dev/rerun-io/rerun/rerun_cpp/tests/component_batch.cpp
In file included from /Users/andreas/dev/rerun-io/rerun/rerun_cpp/tests/component_batch.cpp:3:
In file included from /Users/andreas/dev/rerun-io/rerun/rerun_cpp/src/rerun/archetypes/points2d.hpp:6:
/Users/andreas/dev/rerun-io/rerun/rerun_cpp/src/rerun/archetypes/../component_batch.hpp:38:9: error: static_assert failed due to requirement 'NoAdapterFor<rerun::components::Position2D, std::vector<std::string, std::allocator<std::string>>>::value' "ComponentBatchAdapter is not implemented for this type. It is implemented for for single components as well as std::vector, std::array, and c-arrays of components. You can add your own implementation by specializing ComponentBatchAdapter<TComponent, T> for a given component and your input type T."
        static_assert(
        ^
/Users/andreas/dev/rerun-io/rerun/rerun_cpp/src/rerun/archetypes/../component_batch.hpp:101:52: note: in instantiation of template class 'rerun::ComponentBatchAdapter<rerun::components::Position2D, std::vector<std::string>>' requested here
        ComponentBatch(T&& input) : ComponentBatch(TAdapter<T>()(std::forward<T>(input))) {}
                                                   ^
/Users/andreas/dev/rerun-io/rerun/rerun_cpp/tests/component_batch.cpp:234:39: note: in instantiation of function template specialization 'rerun::ComponentBatch<rerun::components::Position2D>::ComponentBatch<std::vector<std::string> &>' requested here
    rerun::ComponentBatch<Position2D> batch(strings);
```

Another interesting thing about this solution is that it *forced* me to exclude the adapter on T==TComponent because otherwise it made resolution ambiguous. That's nice because that was one of my criticisms: Previously I wasn't sure if we'd take this adapter in those cases, now it's definitely impossible :)

**Review by Wumpf — APPROVED** [review]
happy to ship with my change on top. Let's wait for @jleibs to counter review before landing!

**Review by jleibs — COMMENTED** [review]
Looks great to me!

**Comment by Wumpf** [comment]
cargo deny error is known. 🚢  time!

**Pull request merged**
